accepted/future-releases/static-extension-methods/design-document.md

```diff
@@ -0,0 +1,504 @@
+
+### Extending Static Members
+
+The feature above only extends instance members. There is no way to add a new static member on an existing type, something that should logically be a *simpler* operation.
+
+We could allow
+
+```dart
+extension MyInt on num {
+  int get double => this * s;
+  static int get random => 4;
+}
+```
+
+to introduce both a `double` instance getter on `num` instances and a `random` getter on `num` itself, usable as `var n = num.random;`
+
+However, while this is possible, not all `on` types are class or mixin types. It is not clear what it would mean to put static methods on `on` types of extension like:
+
+- `extension X on Iterable<int>`
+- `extension X<T extends Comparable<T>> on Iterable<T>`
+- `extension X on int Function(int)`
+- `extension X on FutureOr<int>`
+- `extension X on int?`
+
+For the first two, we could put the static members on the `Iterable` class, but since the extension does not apply to *all* iterables, it is not clear that this is correct.
+
+For `int Function(int)` and `FutureOr<int>`, it's unclear how to call such a static method at all. We can denote`int Function(int)` with a type alias, but putting static members on type aliases is a new concept.  We could put the static method on `Function`, but that's not particularly discoverable, and why not require that they are put on `Functon` explicitly. For `FutureOr`, we could allow static members on `FutureOr` (which is a denotable type), but again it seems spurious. For `int?`, we could put the method on `int`, but why not  just require that it's on `int`.
+
+The issue here is that the type patterns used by `on` are much more powerful than what is necessary to put static members on class types.
+
+It would probably be more readable to introduce a proper static extension declaration:
+
+```dart
+static extension Foo on int {  // or: extension Foo on static int 
+  int fromList(List l) => l.length;
+}
+
+...
+  print(int.fromList([1, 2])); // 2
+```
+
+where the `on` type must be something that can already have static methods. 
+
+The disadvantage is that if you want to introduce related functionality that is both static and instance methods on a class, then you need to write two extensions with different names.
+
+If we allow extension static declarations like these, we can also allow extension constructors.
+
+### Explicitly Specify Related Declarations
+
+The above resolution strategy uses an implicit "more specific" relation between applicable extensions to select one of them (when possible). This may cause surprises when two unrelated extensions happen to both apply.
+
+Alternatively, we can allow extensions to declare that they are *related*, and only allow conflicts between related extensions, which are presumably aware of each other.
+
+The syntax would be:
+
+```dart
+extension Foo<T> on Iterable<T> { twizzle() {...} fromp() { ... }}
+extension Bar<T> extends Foo<T> on List<T> { twizzle() { ... } } 
```

Comment:
Do we need "implements" as well?  Otherwise, it feels like I can get stuck:

```
extension Foo on A { twizzle() {...} }
extension Bar on I  { twizzle() { ... } } 
class B extends A implements I {}
extension Baz extends Foo on A { twizzle() {...} } 
B().twizzle() // Error, because Baz doesn't cover I, right?
```

Reply:
Probbly not. Interfaces are useful for generalizing over virtual lookups without having any specific implementation. A static invocation is always a concrete implementation, and is unaffected by interfaces.